File: ham.js
function HAM(machineState, incomingState, currentState, incomingValue, currentValue){
    if(machineState < incomingState){
        return {defer: true};
    }
    if(incomingState < currentState){
        return {historical: true};
    }
    if(currentState < incomingState){
        return {converge: true, incoming: true};
    }
    if(incomingState === currentState){
        incomingValue = Lexical(incomingValue) || "";
        currentValue = Lexical(currentValue) || "";
        if(incomingValue === currentValue){
            return {state: true};
        }
        if(incomingValue < currentValue){
            return {converge: true, current: true};
        }
        if(currentValue < incomingValue){
            return {converge: true, incoming: true};
        }
    }
    return {err: "Invalid CRDT Data: "+ incomingValue +" to "+ currentValue +" at "+ incomingState +" to "+ currentState +"!"};
}
var Lexical = JSON.stringify


HAM.mix = function(change, graph){
    var machine = (+new Date), diff;
    Object.keys(change).forEach(function(soul){
        var node = change[soul];
        Object.keys(node).forEach(function(key){
            var val = node[key];
            if('_' == key){ return }
            var state = node._['>'][key];
            var was = (graph[soul]||{_:{'>':{}}})._['>'][key] || -Infinity;
            var known = (graph[soul]||{})[key];
            var ham = HAM(machine, state, was, val, known);
            if(!ham.incoming){
                if(ham.defer){
                    console.log("DEFER", key, val);
                    // you'd need to implement this yourself.
                }
                return;
            }
            (diff || (diff = {}))[soul] = diff[soul] || {_:{'#':soul, '>':{}}};
            graph[soul] = graph[soul] || {_:{'#':soul, '>':{}}};
            graph[soul][key] = diff[soul][key] = val;
            graph[soul]._['>'][key] = diff[soul]._['>'][key] = state;
        })
    });
    return diff;
}

try{module.exports = HAM}catch(e){}

File: src/ham.rs
use anyhow::{anyhow, Result};
use serde_json::{json, Value};

use crate::util::{timestamp, SOUL, METADATA, STATE};

struct HAM {
	converge: bool,
	current: bool,
	defer: bool,
	historical: bool,
	incoming: bool,
	state: bool,
}

impl Default for HAM {
	fn default() -> Self {
		Self {
			converge: false,
			current: false,
			defer: false,
			historical: false,
			incoming: false,
			state: false,
		}
	}
}

fn ham(
	machine_state: u64,
	incoming_state: u64,
	current_state: u64,
	incoming_value: String,
	current_value: String,
) -> Result<HAM> {
	if machine_state < incoming_state {
        return Ok(HAM { defer: true, ..Default::default() });
    } else if incoming_state < current_state {
        return Ok(HAM { historical: true, ..Default::default() });
    } else if current_state < incoming_state {
        return Ok(HAM { converge: true, incoming: true, ..Default::default() });
    } else if incoming_state == current_state {
        if incoming_value == current_value {
        	return Ok(HAM { state: true, ..Default::default() });
        } else if incoming_value < current_value {
            return Ok(HAM { converge: true, current: true, ..Default::default() });
        } else if current_value < incoming_value {
            return Ok(HAM { converge: true, incoming: true, ..Default::default() });
        }
    }

    Err(anyhow!("Invalid CRDT Data: {} to {} at {} to {}!",
    	incoming_value, current_value, incoming_state, current_state))
}

pub fn mix_ham(change: Value, graph: &mut Value) -> Value {
	let machine = timestamp();
	let mut diff = json!({});

	for (soul, node) in change.as_object().unwrap().iter() {
		for (key, val) in node.as_object().unwrap().iter() {
			if let SOUL | METADATA | STATE = &key[..] {
				continue;
			}

			let state = node[METADATA][STATE][key].as_u64().unwrap();
			let was = match graph.get(soul) {
				Some(node) => match node[METADATA][STATE].get(key) {
					Some(val) => val.as_u64().unwrap(),
					None => 0,
				},
				None => 0,
			};
			let val = val.to_string();
			let known = match graph.get(soul) {
				Some(node) => node[key].to_string(),
				None => "".to_string(),
			};

			let ham = match ham(machine, state, was, val.clone(), known) {
				Ok(ham) => ham,
				Err(err) => {
					log::error!("{}", err);
					continue;
				},
			};

			if !ham.incoming {
				if ham.defer {
					log::info!("DEFER {} {}", key, val);
				}
				continue;
			}

			if graph[soul].is_null() {
				graph[soul] = json!({METADATA:{SOUL:soul, STATE:{}}});
			}

			if diff[soul].is_null() {
				diff[soul] = json!({METADATA:{SOUL:soul, STATE:{}}});
			}

			graph[soul][key] = json!(val);
			diff[soul][key] = json!(val);

			graph[soul][METADATA][STATE][key] = json!(state);
			diff[soul][METADATA][STATE][key] = json!(state);
   		}
	}

	diff
}
